File: client2.c
#include <stdlib.h>
#include <stdio.h>
#include <dlfcn.h>

// type def
typedef long long go_int;
typedef double go_float64;
typedef struct{void *arr; go_int len; go_int cap;} go_slice;
typedef struct{const char *p; go_int len;} go_str;

int main(int argc, char **argv) {
    void *handle;
    char *error;

    // use dlopen to load shared object
    handle = dlopen ("awesomelib/awesome.so", RTLD_LAZY);
    if (!handle) {
        fputs (dlerror(), stderr);
        exit(1);
    }
    
    // lookup function symbols by name
    // and call the functions

    // Call  Add()
    go_int (*add)(go_int, go_int)  = dlsym(handle, "Add");
    if ((error = dlerror()) != NULL)  {
        fputs(error, stderr);
        exit(1);
    }
    go_int sum = (*add)(12, 99);
    printf("awesome.Add(12, 99) = %d\n", sum);

    // Call  Cosine()
    go_float64 (*cosine)(go_float64) = dlsym(handle, "Cosine");
    if ((error = dlerror()) != NULL)  {
        fputs(error, stderr);
        exit(1);
    }
    go_float64 cos = (*cosine)(1.0);
    printf("awesome.Cosine(1) = %f\n", cos);

    // Call Sort
    void (*sort)(go_slice) = dlsym(handle, "Sort");
    if ((error = dlerror()) != NULL)  {
        fputs(error, stderr);
        exit(1);
    }
    go_int data[5] = {44,23,7,66,2};
    go_slice nums = {data, 5, 5};
    sort(nums);
    printf("awesome.Sort(44,23,7,66,2): ");
    for (int i = 0; i < 5; i++){
        printf("%d,", ((go_int *)data)[i]);
    }
    printf("\n");

    // Call Log
    go_int (*log)(go_str) = dlsym(handle, "Log");
    if ((error = dlerror()) != NULL)  {
        fputs(error, stderr);
        exit(1);
    }
    go_str msg = {"Hello from C!", 13};
    log(msg);
    
    // close file handle when done
    dlclose(handle);
}

File: client2.zig
const std = @import("std");

const go_int = i64;
const go_float = f64;
const go_str = []const u8;
const go_slice = extern struct {
    data: *[]go_int,
    len: go_int,
    cap: go_int,
};
var Add: *const fn (a: go_int, b: go_int) go_int = undefined;
var Cosine: *const fn (n: go_float) go_float = undefined;
var SortPtr: *const fn (nums: *go_slice) void = undefined;
var Log: *const fn (str: go_str) go_int = undefined;

pub fn main() !void {
    // use dlopen to load shared object
    var awesome = try std.DynLib.open("./awesome.so");
    // close file handle when done
    defer awesome.close();

    // resolve Add symbol and assign to fn ptr
    Add = awesome.lookup(@TypeOf(Add), "Add") orelse return error.LookupFail;
    // call Add()
    std.debug.print("awesome.Add(12,99) = {d}\n", .{Add(12, 99)});

    // resolve Cosine symbol
    Cosine = awesome.lookup(@TypeOf(Cosine), "Cosine") orelse return error.LookupFail;
    // Call Cosine
    std.debug.print("awesome.Cosine(1) = {d}\n", .{Cosine(1.0)});

    // resolve SortPtr symbol
    SortPtr = awesome.lookup(@TypeOf(SortPtr), "SortPtr") orelse return error.LookupFail;
    const data = [5]go_int{ 44, 23, 7, 66, 2 };
    const nums = go_slice{ .data = @ptrCast(@constCast(&data)), .len = data.len, .cap = data.len };
    // call SortPtr
    SortPtr(@ptrCast(@constCast(&nums)));
    std.debug.print("awesome.SortPtr(44,23,7,66,2): ", .{});
    for (@as(*align(1) const [nums.len]go_int, @ptrCast(nums.data))) |n| {
        std.debug.print("{d},", .{n});
    }
    std.debug.print("\n", .{});

    // resolve Log symbol
    Log = awesome.lookup(@TypeOf(Log), "Log") orelse return error.LookupFail;
    // call Log
    _ = Log("Hello from Zig!");
}
